sensor/common/registry/lazy_tls_image_registry.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+// LazyTLSCheckRegistry is a wrapper around ImageRegistry that
+// will perform a TLS check on first use instead of
+// expecting a TLS check to have been done prior.
+type LazyTLSCheckRegistry struct {
+	// source represents the registry to create, however
+	// the insecure flag associated with the underlying config is
+	// assumed to be uninitialized.
+	source *storage.ImageIntegration
+
+	// creater is responsible for creating the underlying registry
+	// based on imageIntegration when ready to do so.
+	creator types.Creator
+
+	dataSource *storage.DataSource
+
+	// once is used to control the lazy initialization of the underlying
+	// registry
+	once sync.Once
+
+	// initError will be populated if initialization of the underlying
+	// registry fails.  If non nil will be returned or will alter
+	// the return of the implemented ImageRegistry methods.
+	initError error
+
+	// checkTLSFunc is the function used to lazily check the TLS support
+	// of the registry.
+	tlsCheckCache *tlsCheckCacheImpl
+
+	// registry is the underlying registry that is only populated after
+	// successful lazy initialization.
+	registry types.Registry
+}
+
+var _ types.ImageRegistry = (*LazyTLSCheckRegistry)(nil)
+
+func (l *LazyTLSCheckRegistry) Config() *types.Config {
+	// TODO: Find better way (in registry store) to determine if a hostname matches so
+	// that initialization is NOT performed here and can be deferred to the first Metadata
+	// call (ideal).
+	// May have to dupe this logic:
+	// https://github.com/stackrox/stackrox/blob/cb378f5d1341323c887d07868598d64a8a82d214/pkg/registries/docker/docker.go#L79
+	l.lazyInit()
+	if l.initError != nil {
+		log.Debugf("Returning nil config for %q due to init error: %v", l.source.GetId(), l.initError)
+		return nil
+	}
+
+	return l.registry.Config()
+}
+
+func (l *LazyTLSCheckRegistry) DataSource() *storage.DataSource {
+	return l.dataSource
+}
+
+func (l *LazyTLSCheckRegistry) HTTPClient() *http.Client {
+	l.lazyInit()
+	if l.initError != nil {
+		log.Debugf("Returning nil http client for %q due to init error: %v", l.source.GetId(), l.initError)
+		return nil
+	}
+
+	return l.registry.HTTPClient()
+}
+
+func (l *LazyTLSCheckRegistry) Match(image *storage.ImageName) bool {
+	l.lazyInit()
+	if l.initError != nil {
+		return false
+	}
+
+	return l.registry.Match(image)
+}
+
+func (l *LazyTLSCheckRegistry) Metadata(image *storage.Image) (*storage.ImageMetadata, error) {
+	l.lazyInit()
+	if l.initError != nil {
+		return nil, l.initError
+	}
+
+	return l.registry.Metadata(image)
+}
+
+func (l *LazyTLSCheckRegistry) Name() string {
+	return l.source.GetName()
+}
+
+func (l *LazyTLSCheckRegistry) Source() *storage.ImageIntegration {
+	l.lazyInit()
+	return l.source
+}
+
+func (l *LazyTLSCheckRegistry) Test() error {
+	l.lazyInit()
+	if l.initError != nil {
+		return l.initError
+	}
+
+	return l.registry.Test()
+}
+
+func (l *LazyTLSCheckRegistry) lazyInit() {
+	l.once.Do(func() {
```

Comment:
Would it make sense to retry this the next time it is called if for whatever reason we fail the first time (e.g. timeout)? We still would have the cached value, so if different calls happen while the cache is still valid then we won't need to worry about multiple calls to `ChackTLS`. I know that in the current state of the code we will not do this unless there's an update in the secret, but maybe it's something that we can improve here. WDYT?

Reply:
Great idea, will work that in.